pl_examples/domain_templates/reinforce_learn_ppo.py

```diff
@@ -0,0 +1,474 @@
+
+        elif optimizer_idx == 1:
+            loss_critic = self.critic_loss(state, action, old_logp, qval, adv)
+            self.log('loss_critic', loss_critic, on_step=False, on_epoch=True, prog_bar=False, logger=True)
+
+            return loss_critic
+
+    def configure_optimizers(self) -> List[Optimizer]:
+        """ Initialize Adam optimizer"""
+        optimizer_actor = optim.Adam(self.actor.parameters(), lr=self.lr_actor)
+        optimizer_critic = optim.Adam(self.critic.parameters(), lr=self.lr_critic)
+
+        return optimizer_actor, optimizer_critic
+
+    def optimizer_step(self, *args, **kwargs):
+        """
+        Run 'nb_optim_iters' number of iterations of gradient descent on actor and critic
+        for each data sample.
+        """
+        for i in range(self.nb_optim_iters):
+            super().optimizer_step(*args, **kwargs)
+
+    def _dataloader(self) -> DataLoader:
+        """Initialize the Replay Buffer dataset used for retrieving experiences"""
+        dataset = ExperienceSourceDataset(self.train_batch)
+        dataloader = DataLoader(dataset=dataset, batch_size=self.batch_size)
+        return dataloader
+
+    def train_dataloader(self) -> DataLoader:
+        """Get train loader"""
+        return self._dataloader()
+
+    @staticmethod
+    def add_model_specific_args(parent_parser):  # pragma: no-cover
+        parser = argparse.ArgumentParser(parents=[parent_parser])
+        parser.add_argument("--env", type=str, default="CartPole-v0")
+        parser.add_argument("--gamma", type=float, default=0.99, help="discount factor")
+        parser.add_argument("--lam", type=float, default=0.95, help="advantage discount factor")
+        parser.add_argument("--lr_actor", type=float, default=3e-4, help="learning rate of actor network")
+        parser.add_argument("--lr_critic", type=float, default=1e-3, help="learning rate of critic network")
+        parser.add_argument("--max_episode_len", type=int, default=1000, help="capacity of the replay buffer")
+        parser.add_argument("--batch_size", type=int, default=512, help="batch_size when training network")
+        parser.add_argument("--steps_per_epoch", type=int, default=2048,
+                            help="how many action-state pairs to rollout for trajectory collection per epoch")
+        parser.add_argument("--nb_optim_iters", type=int, default=4,
+                            help="how many steps of gradient descent to perform on each batch")
+        parser.add_argument("--clip_ratio", type=float, default=0.2,
+                            help="hyperparameter for clipping in the policy objective")
+
+        return parser
+
+
+def main(args) -> None:
+    model = PPOLightning(**vars(args))
+
+    trainer = pl.Trainer(max_epochs=70)
+    trainer.fit(model)
+
+
+if __name__ == '__main__':
+    cli_lightning_logo()
+    torch.manual_seed(0)
+    np.random.seed(0)
+
+    parser = argparse.ArgumentParser(add_help=False)
```

Comment:
I don't think it works with DDP- but I'm saying that only because all the RL [implementations](https://github.com/PyTorchLightning/pytorch-lightning-bolts/tree/master/pl_bolts/models/rl) in lightning bolts don't support it (says in their comments). And this PPO implementation has very similar structure and data generation as [Reinforce](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/8a4cf8f61644c28d6df54ccffe3a52d6f5fce5a6/pl_bolts/models/rl/reinforce_model.py#L165)

An iterable dataset is generated before every epoch by `train_batch` method, which performs trajectory collection using the model